### PR TITLE
[low-power] add cli command `macsend` for certification

### PR DESCRIPTION
**Pull request by Irving-cl**
This command allows an Rx-Off-When-Idle device to send an empty mac data frame or mac datarequest to its parent. This command is only used for certification.

### CODE REVIEW

File: tests/scripts/expect/v1_2-sed_mac_send.exp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+set spawn_id [spawn_node 2]
+send "panid 0x1234\n"
+expect "Done"
+send "ifconfig up\n"
+expect "Done"
+send "mode s\n"
+expect "Done"
+send "thread start\n"
+expect "Done"
+wait_for "state" "child"
```
Comment (Moandor-y) [inline tests/scripts/expect/v1_2-sed_mac_send.exp:50]: How about setting up the nodes using `setup_nodes` function?

File: tests/scripts/expect/v1_2-sed_mac_send.exp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+send "macsend data\n"
+expect "Done"
+send "macsend datarequest\n"
+expect "Done"
+
```
Comment (Moandor-y) [inline tests/scripts/expect/v1_2-sed_mac_send.exp:55]: Nodes created by `spawn_node` should be disposed by calling `dispose`.
The same for `setup_leader` and `dispose_leader`, `setup_nodes` and `dispose_nodes`. They should be called pairwise.

File: src/core/thread/mesh_forwarder.cpp [hunk 3]
```diff
@@ -1281,6 +1293,71 @@ otError MeshForwarder::GetFramePriority(const uint8_t *     aFrame,
...
+exit:
+    if (error != OT_ERROR_NONE && message != nullptr)
+    {
+        message->Free();
+    }
+    otLogDebgMac("Send empty message, error:%d", error);
+    return error;
+}
+
+void MeshForwarder::PrepareEmptyDataFrame(Mac::TxFrame &aFrame)
```
Comment (librasungirl) [inline src/core/thread/mesh_forwarder.cpp:1320]: It is quite similar to existing `IndirectSender::PrepareEmptyFrame(Mac::TxFrame &aFrame, Child &aChild, bool aAckRequest)` , better to have a common function to avoid redundancy?

File: tests/scripts/expect/_multinode.exp [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 source "tests/scripts/expect/_common.exp"
 
-proc setup_nodes {} {
+proc setup_nodes {{childmode {rs}}} {
```
Comment (librasungirl) [inline tests/scripts/expect/_multinode.exp:32]: note: bit `s` is being removed in https://github.com/openthread/openthread/pull/5560

File: tests/scripts/expect/v1_2-sed_mac_send.exp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+source "tests/scripts/expect/_common.exp"
+source "tests/scripts/expect/_multinode.exp"
+
+setup_nodes "s"
```
Comment (librasungirl) [inline tests/scripts/expect/v1_2-sed_mac_send.exp:33]: note: bit `s` is being removed in #5560

**Review by Moandor-y — APPROVED** [review]
Good for exp

File: src/core/thread/mesh_forwarder.cpp [hunk 6]
```diff
@@ -1281,6 +1293,71 @@ otError MeshForwarder::GetFramePriority(const uint8_t *     aFrame,
...
+    message = Get<MessagePool>().New(Message::kTypeMacEmptyData, 0);
+    VerifyOrExit(message != nullptr, error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error = SendMessage(*message));
+
+exit:
+    if (error != OT_ERROR_NONE && message != nullptr)
+    {
+        message->Free();
+    }
```
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp:1315]:
```suggestion
    FreeMessageOnError(message, error)
```

File: src/core/thread/mesh_forwarder.cpp (hunk 3 above)
Comment (abtink) [reply src/core/thread/mesh_forwarder.cpp:1320]: Was about to suggest the same thing. I fully agree. Should re-use the method.

File: src/core/thread/indirect_sender.cpp [hunk 7]
```diff
@@ -418,37 +418,56 @@ uint16_t IndirectSender::PrepareDataFrame(Mac::TxFrame &aFrame, Child &aChild, M
...
     }
 
-    fcf |= (macDest.IsShort()) ? Mac::Frame::kFcfDstAddrShort : Mac::Frame::kFcfDstAddrExt;
+    fcf |= (aMacDest.IsShort()) ? Mac::Frame::kFcfDstAddrShort : Mac::Frame::kFcfDstAddrExt;
     fcf |= (macSource.IsShort()) ? Mac::Frame::kFcfSrcAddrShort : Mac::Frame::kFcfSrcAddrExt;
+    {
+        FILE *p = fopen("debug.log", "a+");
+        fprintf(p, "dest short:%d, src short:%d\n", aMacDest.IsShort(), macSource.IsShort());
+        fclose(p);
+    }
```
Comment (abtink) [inline src/core/thread/indirect_sender.cpp:451]:
```
    {
         FILE *p = fopen("debug.log", "a+");
         fprintf(p, "dest short:%d, src short:%d\n", aMacDest.IsShort(), macSource.IsShort());
         fclose(p);
     }
```

What is intention behind this block? 
(is it added for testing only - if so please remove or use debug log instead)

File: src/core/thread/indirect_sender.cpp [hunk 8]
```diff
@@ -418,37 +418,56 @@ uint16_t IndirectSender::PrepareDataFrame(Mac::TxFrame &aFrame, Child &aChild, M
...
 {
-    uint16_t     fcf;
-    Mac::Address macSource, macDest;
-
+    Mac::Address macDest;
     aChild.GetMacAddress(macDest);
+    PrepareEmptyFrame(aFrame, macDest, aAckRequest);
+}
 
+void IndirectSender::PrepareEmptyFrame(Mac::TxFrame &aFrame, Mac::Address &aMacDest, bool aAckRequest)
```
Comment (abtink) [inline src/core/thread/indirect_sender.cpp:426]: Use `const &` ?
```suggestion
void IndirectSender::PrepareEmptyFrame(Mac::TxFrame &aFrame, const Mac::Address &aMacDest, bool aAckRequest)
```

File: src/core/thread/indirect_sender.hpp [hunk 9]
```diff
@@ -222,6 +223,7 @@ class IndirectSender : public InstanceLocator, public IndirectSenderBase
     void     RequestMessageUpdate(Child &aChild);
     uint16_t PrepareDataFrame(Mac::TxFrame &aFrame, Child &aChild, Message &aMessage);
     void     PrepareEmptyFrame(Mac::TxFrame &aFrame, Child &aChild, bool aAckRequest);
+    void     PrepareEmptyFrame(Mac::TxFrame &aFrame, Mac::Address &aDstAddr, bool aAckRequest);
```
Comment (abtink) [inline src/core/thread/indirect_sender.hpp:226]:
```suggestion
    void     PrepareEmptyFrame(Mac::TxFrame &aFrame, const Mac::Address &aDstAddr, bool aAckRequest);
```

File: src/cli/cli.cpp [hunk 10]
```diff
@@ -4134,6 +4140,26 @@ otError Interpreter::ProcessMacRetries(uint8_t aArgsLength, char *aArgs[])
     return error;
 }
 
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+otError Interpreter::ProcessMacSend(uint8_t aArgsLength, char *aArgs[])
+{
+    otError error = OT_ERROR_INVALID_ARGS;
+    VerifyOrExit(aArgsLength == 1, OT_NOOP);
```
Comment (abtink) [inline src/cli/cli.cpp:4147]: newline after variable declaration?

File: src/core/thread/mesh_forwarder.cpp [hunk 11]
```diff
@@ -460,6 +466,16 @@ otError MeshForwarder::HandleFrameRequest(Mac::TxFrame &aFrame)
         SendMesh(*mSendMessage, aFrame);
         break;
 
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    case Message::kTypeMacEmptyData:
+    {
+        Mac::Address macDestAddr;
+        macDestAddr.SetExtended(Get<Mle::MleRouter>().GetParent().GetExtAddress());
```
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp:473]: Is it required to use extended address of parent? Can we use short address?

**Review by abtink — COMMENTED** [review]
Looks good overall. 
Few suggestions/questions below. Thanks.

**Review by abtink — APPROVED** [review]
LGTM. Thanks.

File: src/core/thread/mesh_forwarder.cpp [hunk 12]
```diff
@@ -460,6 +466,16 @@ otError MeshForwarder::HandleFrameRequest(Mac::TxFrame &aFrame)
...
 
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    case Message::kTypeMacEmptyData:
+    {
+        Mac::Address macDestAddr;
+        macDestAddr.SetShort(Get<Mle::MleRouter>().GetParent().GetRloc16());
+        Get<IndirectSender>().PrepareEmptyFrame(aFrame, macDestAddr, /* aAckRequest */ false);
+    }
+    break;
+#endif
```
Comment (librasungirl) [inline src/core/thread/mesh_forwarder.cpp:477]: Move out of `#if OPENTHREAD_FTD` scope?

File: tests/scripts/thread-cert/v1_2_test_csl_transmission.py [hunk 13]
```diff
@@ -71,6 +71,11 @@ def test(self):
         self.assertTrue(self.nodes[LEADER].ping(self.nodes[SSED_1].get_rloc()))
         self.simulator.go(5)
 
+        self.nodes[SSED_1].send_mac_data()
+        self.simulator.go(1)
+        self.nodes[SSED_1].send_mac_datarequest()
+        self.simulator.go(1)
+
```
Comment (librasungirl) [inline tests/scripts/thread-cert/v1_2_test_csl_transmission.py:78]: For SSED_1 topology config in above Line#48-#51, add `is_mtd: True` to use ot-cli-mtd and set `USE_MTD=1`  env.

File: tests/scripts/thread-cert/v1_2_test_csl_transmission.py [hunk 14]
```diff
@@ -71,6 +71,11 @@ def test(self):
         self.assertTrue(self.nodes[LEADER].ping(self.nodes[SSED_1].get_rloc()))
         self.simulator.go(5)
 
+        self.nodes[SSED_1].send_mac_data()
+        self.simulator.go(1)
+        self.nodes[SSED_1].send_mac_datarequest()
+        self.simulator.go(1)
```
[the remark this answers is not in the document]
Comment (librasungirl) [reply tests/scripts/thread-cert/v1_2_test_csl_transmission.py:77]: Better to add traffic check, as it skips the error when test SSED of `ot-cli-mtd` which didn't send empty data frame.

File: src/core/thread/mesh_forwarder.cpp (hunk 12 above)
Comment (librasungirl) [reply src/core/thread/mesh_forwarder.cpp:477]: should this unicast packet request ack?

File: src/core/thread/indirect_sender.cpp [hunk 15]
```diff
@@ -418,37 +418,51 @@ uint16_t IndirectSender::PrepareDataFrame(Mac::TxFrame &aFrame, Child &aChild, M
...
-    IgnoreError(aFrame.SetSrcPanId(Get<Mac::Mac>().GetPanId()));
-    aFrame.SetDstAddr(macDest);
+    if (aFrame.IsDstPanIdPresent())
+    {
+        aFrame.SetDstPanId(Get<Mac::Mac>().GetPanId());
+    }
+    if (aFrame.IsSrcPanIdPresent())
+    {
+        IgnoreError(aFrame.SetSrcPanId(Get<Mac::Mac>().GetPanId()));
+    }
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/core/thread/indirect_sender.cpp:458]: Yes. In `SetSrcPanId` there's an assertion.

File: tests/scripts/thread-cert/v1_2_test_csl_transmission.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Irving-cl) [reply tests/scripts/thread-cert/v1_2_test_csl_transmission.py]: The test code is just to cover the new code. As the new code are already covered by the tests in expect, I removed the lines here.

File: src/core/thread/mesh_forwarder.cpp (hunk 12 above)
Comment (Irving-cl) [reply src/core/thread/mesh_forwarder.cpp]: Note that we are now reusing the code from `IndirectSender`which only exists in FTD.

File: src/core/thread/indirect_sender.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (librasungirl) [reply src/core/thread/indirect_sender.cpp]: I can see there is assert in `SetDstPanId`, but I fails to see it in `SetSrcPanId` in the master upstream.

[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/core/thread/indirect_sender.cpp]: Before this PR, the empty data frame always uses short address so the assertion won't be triggered. I found the problem because I used Ext address for empty data frame in early time in this PR. Though I changed to use short address now, it's necessary to add the check (if we use this method again with extended address in the future).

**Review by librasungirl — APPROVED** [review]

File: include/openthread/link.h [hunk 16]
```diff
@@ -1109,6 +1109,20 @@ otError otLinkSetEnabled(otInstance *aInstance, bool aEnable);
  */
 bool otLinkIsEnabled(otInstance *aInstance);
 
+/**
+ * This function instructs the device to send an empty IEEE 802.15.4 data frame.
+ *
+ * This function is only supported on an Rx-Off-When-Idle device to send an empty data frame to its parent.
```
Comment (jwhui) [inline include/openthread/link.h:1115]: Add:
```
Note: available only when `OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE` is enabled.
```

File: src/cli/README.md [hunk 17]
```diff
@@ -2047,6 +2047,20 @@ Set the number of indirect TX retries on the MAC layer.
 Done
 ```
 
+### mac send \<op\>
```
Comment (simonlingoogle) [inline src/cli/README.md:2050]: nit: can we merge this command with `fake` command which is currently used to send fake `/a/an` for certification purpose.  
e.g. 
```bash
> fake datarequest
Done
> fake emptydata
Done
```
@librasungirl @Irving-cl

File: src/cli/README.md [hunk 18]
```diff
@@ -2047,6 +2047,20 @@ Set the number of indirect TX retries on the MAC layer.
...
+
+Instruct an Rx-Off-When-Idle device to send a mac frame to its parent. The mac frame could be either a mac data request or an empty mac data frame. Use `datarequest` to send a mac data request and `data` to send an empty mac data. This feature is for certification, it can only be used when `OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE` is enabled.
+
+```bash
+> mac send datarequest
+Done
+```
+
+```bash
+> mac send data
```
Comment (simonlingoogle) [inline src/cli/README.md:2060]:
```suggestion
> mac send emptydata
```
nit: be more explicit for empty data frame?

File: src/core/thread/mesh_forwarder.cpp [hunk 19]
```diff
@@ -454,6 +500,16 @@ otError MeshForwarder::HandleFrameRequest(Mac::TxFrame &aFrame)
         OT_ASSERT(aFrame.GetLength() != 7);
         break;
 
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    case Message::kTypeMacEmptyData:
+    {
+        Mac::Address macDestAddr;
+        macDestAddr.SetShort(Get<Mle::MleRouter>().GetParent().GetRloc16());
```
Comment (simonlingoogle) [inline src/core/thread/mesh_forwarder.cpp:507]:
```suggestion
        Mac::Address macDestAddr;

        macDestAddr.SetShort(Get<Mle::MleRouter>().GetParent().GetRloc16());
```
nit: empty line after variable declaration.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 
with nits

File: src/cli/README.md (hunk 17 above)
Comment (Irving-cl) [reply src/cli/README.md:2050]: No strong preference on this. I think we can keep it as it is now since `fake` command sends a IP message and needs a IP address while `mac send` sends a Link layer message and always sends to the parent. But we can consider again in the future if we need a specific command to send these kind of messages when there's more of them.

**Review by jwhui — APPROVED** [review]
